core/src/ldml/ldml_processor.cpp

```diff
@@ -59,23 +59,130 @@ namespace {
   };
 }
 
+/**
+ * @brief Usage: `KMXPLUS_PRINTF(("str: %s\n", "something"));`
+ * Note double parens
+ * \def KMXPLUS_DEBUG
+ */
+#if KMXPLUS_DEBUG
+#include <iostream>
+#define KMXPLUS_NOTEQUAL(expect, actual) std::cerr << __FILE__ << ":" << __LINE__ << ": ASSERT FAILED: " \
+              << #actual << "=" << (actual) << ", expected " << (expect) << std::endl
+#define KMXPLUS_PRINTLN(msg) std::cerr  << __FILE__ << ":" << __LINE__ << ": " msg << std::endl;
+#else
+#define KMXPLUS_NOTEQUAL(expect, actual)
+#define KMXPLUS_PRINTLN(msg)
+#endif
+
 namespace km {
 namespace kbp {
 
 ldml_processor::ldml_processor(path const & kb_path, const std::vector<uint8_t> &data)
 : abstract_processor(
     keyboard_attributes(kb_path.stem(), KM_KBP_LMDL_PROCESSOR_VERSION, kb_path.parent(), {})
-  ), _valid(false), rawdata(data) // TODO-LDML: parse the data, don't just copy it
+  ), _valid(false), vkey_to_string()
 {
+
+/**
+ * Assert something, just in this function
+ * If fails, print a message and exit as invalid
+ * \def KMXPLUS_ASSERT
+ */
+#define KMXPLUS_ASSERT(expect,actual) if ((expect) != (actual)) { \
+    KMXPLUS_NOTEQUAL(expect, actual); \
+    _valid = false; \
+    return; \
+  }
+  const kmx::COMP_KMXPLUS_SECT *sect = nullptr;
+  const kmx::COMP_KMXPLUS_STRS *strs = nullptr;
+  const kmx::COMP_KMXPLUS_KEYS *keys = nullptr;
+
   // TODO-LDML: load the file from the buffer (KMXPlus format)
   // Note: kb_path is essentially debug metadata here
-  assert(data.size() != 0);
+  KMXPLUS_ASSERT(true, data.size() > sizeof(kmx::COMP_KEYBOARD_EX));
 
+  // Locate the structs here, but still retain ptrs to the raw structs.
   const kmx::PCOMP_KEYBOARD comp_keyboard = (kmx::PCOMP_KEYBOARD)data.data();
+  KMXPLUS_ASSERT(true, !!(comp_keyboard->dwFlags & KF_KMXPLUS));
+  const kmx::COMP_KEYBOARD_EX* ex = reinterpret_cast<const kmx::COMP_KEYBOARD_EX*>(comp_keyboard);
+
+  // validate size and offset
+  KMXPLUS_ASSERT(true, ex->kmxplus.dwKMXPlusSize >  LDML_LENGTH_SECT);
```

Comment:
What if there is only a `SECT` section? Technically structurally valid? This assumption may need discussion -- it impacts other parts of the keyboard load below.

```suggestion
  KMXPLUS_ASSERT(true, ex->kmxplus.dwKMXPlusSize >= LDML_LENGTH_SECT);
```

Reply:
I'm not sure that makes sense.  It's a keyboard, so it should have keys, right? at least a 0-length array of keys.  If it was a generic container that had other content, then perhaps `keys` would be optional.

Reply:
A "null" keyboard is a valid keyboard -- useful if nothing else for unit tests. But also we may well see keyboards that use a US English layout but have transforms. So I do think I can see a case for skipping a `keys` section.

In any case, I think we want to be clear on what is structurally valid vs functional validity. Then, we need to clearly document which sections are required in the kmxplus spec file.

Reply:
Sounds good to me then.  No keys, just transforms. 

- [ ] update the spec to clarify validity